docs/contracts-intro.md

```diff
@@ -2,11 +2,181 @@
+relating to a particular clinical data domain. OpenSAFELY Contracts are
+authored by the OpenSAFELY platform developers: the Contract designs are
+informed by discussion and feedback from data providers and researchers.
+
+Researchers write dataset definitions to access data tables from
+OpenSAFELY backends. All available data tables within OpenSAFELY satisfy
+an OpenSAFELY Contract, regardless of which data provider operates the
+backend.
+
+TODO: does a table only satisfy one contract, or possibly multiple?
+
+## The OpenSAFELY Contracts reference
+
+The [OpenSAFELY Contracts reference](contracts-reference.md) has full
+details of each individual Contract. When writing a dataset definition,
+researchers should consult [the OpenSAFELY Contracts
+reference](contracts-reference.md) to see:
+
+* the data available in OpenSAFELY backends
+* how the data is structured
+* what that data represents
+
+!!! warning
+    OpenSAFELY Contracts apply only when accessing OpenSAFELY data using
+    *dataset definitions*, written to run with Data Builder. Contracts
+    do not apply to *study definitions* written for the existing cohort
+    extractor.
+
+## The goals of OpenSAFELY Contracts
+
+Contracts make it easier for:
+
+* *Researchers* to understand what data is available in
+  OpenSAFELY and how that data is represented.
+* *Data providers* to structure their data for OpenSAFELY.
+* *OpenSAFELY's developers* to extend OpenSAFELY to additional
+  healthcare and clinical data domains in future.
+
+!!! note
+    OpenSAFELY Contracts are a new concept still being explored.
+
+    We [value
+    feedback](https://github.com/opensafely/documentation/discussions)
+    from both OpenSAFELY users and data providers, and conversations on
+    how we might further refine the design of individual OpenSAFELY
+    Contracts or the system as a whole.
+
+### The benefits of OpenSAFELY Contracts for researchers
+
+With OpenSAFELY Contracts, researchers benefit from:
+
+* Guarantees on the structure of the data table being queried.
+* A simpler query interface to multiple backends with the ability to:
+    * look up which backends provide the data of interest
+    * verify which backends are compatible with their dataset definitions
+    * run data definitions on multiple different backends, without
+      having to write specific code for each individual backend
+* More consistent documentation that:
+    * describes the commonalities of the data offered by different backends
+    * notes where individual backends provide data that deviates from others
+
+## Common and Custom Contracts
+
+Data backends that are accessible via OpenSAFELY provide data tables. Each
+data table satisfies an OpenSAFELY Contract.
+
+There are two types of Contracts:
+
+* **Common Contracts**, required by all OpenSAFELY backends and covering
+  core clinical data domains
+* **Custom Contracts**, offered by just some backends and covering more
+  specialised clinical data domains
```

Comment:
Is this correct. I'm not sure that we require _all_ backends to implement _all_ common contracts.

I thought the distinction was that common contracts are defined by us and if you want to implement them you have to do so completely and meet all the constraints; whereas custom contracts are defined by a backend provider to cover some data or format specific to their backend.

Reply:
Yes, I think you're right, and this is my misreading.

Reply:
I think our (my?) understanding of the classification of contracts has shifted since we first came up with this classification. I think the idea of "required" contracts is unecessary: we will have tooling for people to see which backends support the concepts *they're* interested in.

I think I now think of "common contracts" as ones that have been validated as expressing concepts that can be applied across two or more backends.  "custom contracts" are ones that have been implemented for only one backend.

It's possible even this distinction isn't particularly useful. Every time a backend wants to join opensafely, we will have a conversation with them about the contracts they will implement, and this conversation may cause us to bump the version of a contract already used by another backend.

On the other hand, we may want to make the fact we've never thought about the generalisability of a new contract really clear through categorisation, or namespacing (so if there's no 111Calls core contract, then the first time we implement it, we would choose "Graphnet/111Calls"; the second time, we might create a new "NHSE/111Calls" contract based on the Graphnet one, and perhaps somehow deprecate the Graphnet one.

I feel this is still not decided and I'd appreciate more feedback, more widely. For now perhaps we should weaken the classification here, and just explain that if a contract has not been validated as generalisable across multiple backends, it will be namespaced as such?

Reply:
> I think I now think of "common contracts" as ones that have been validated as expressing concepts that can be applied across two or more backends. "custom contracts" are ones that have been implemented for only one backend.

This is how I understood it from re-reading. It's my half-remembering, along with not re-reading properly the first time that caused the initial confusion here. 

***

I had some related thoughts that don't really go anywhere, but will note them.

What does including "Custom Contracts" give you over not having it?

* It marks to researchers clearly which contracts are only supported by a single backend.
  * Counterpoint: let's say, in a ridiculous example, that we have 100 backends. Some of the Contracts might be "Custom" and only supported by one backend. Others might be "Common", but only supported by two backends. Is the extra concept and distinction useful here?
* We're trying to encourage data providers to consider structuring their data, where possible, to make things easier for researchers. Having the distinction means they can still provide additional data, or their idiosyncratic forms of existing data. But we're still clearly saying: "this is a special case".
* Likewise, if we're encouraging re-use of existing Contracts reduces our burden in having to concoct ad-hoc contracts for every individual backend.

To me, the distinction feels more related to communication and advocacy, and how you want to position that. And less so a technical decision (but I'm happy to be corrected).

***

All the above said, I think you could do the same by the Contract's name or namespace, if required. That makes it clearer to the person using such a Contract, avoiding needing some kind of separate label. 

And, as we haven't fully decided, for now, maybe it's worth de-emphasising the distinction and avoid introducing a separate concept. Instead:

* warn that some Contracts may be particularly specific to one backend
* warn that such Contracts will be labelled as such, somehow (to be determined)